Define behaviour for empty expressions

YehorBoiar opened the pull request:
This PR is based on the Chris's and Oz's comment on Teams:

"
Chris:
I'd get existing PRs merged in first, but then it might be worth making a pass over empty constraints, I think (please check essence semantics  , and I'm not sure which are implemented yet):
 
and([]) = true
or([]) = false
sum([]) = 0
product([]) = 1
 
it's worth getting these right because (from experience), while no-one is likely to write these, they often end up getting created as parts of more interesting expressions.
 
 Oz:
min([]) and max([]) are a bit different, as min([])=x is always false, but this falseness will arise naturally from the fact they will end up turning into an 'and([])' and an 'or([])', and the or will be false, so there is no need to special-case empty min([]) and max([]) (hopefully!)
 
we can define an abstraction which has a zero vale and an append operator for these.
and is true, /\
or is false, \/
sum is 0, +
product is 1, *
regarding min and max, we already did a bunch of definedness work with Felix last semester (Soph and Yehor taking over this now)
 "

ozgurakgun commented:
we should also actually remove Nothing, shouldn't we?

YehorBoiar commented:
Will do

ozgurakgun commented on crates/conjure_core/src/rules/base.rs:
````diff
@@ -77,11 +77,16 @@ fn remove_nothings(expr: &Expr, _: &Model) -> ApplicationResult {
+/// X([]) ~> Nothing
+/// ```
 #[register_rule(("Base", 8800))]
 fn remove_empty_expression(expr: &Expr, _: &Model) -> ApplicationResult {
````
I just noticed: why is this a separate rule instead of being handled by the full evaluator?

ozgurakgun replied:
@niklasdewally @YehorBoiar @lixitrixi

niklasdewally replied:
By the look of git blame, seems to just be because Nothing was added before the evaluator.

 I agree that it should be an evaluation step.

YehorBoiar replied:
bruh

ozgurakgun replied:
don't know what you mean by "We should just define an empty expression behaviour in rules"

for context, full evaluator here: https://github.com/conjure-cp/conjure-oxide/blob/dc7f98bd5cdf32122de3ce4fb8235d86747d4b88/crates/conjure_core/src/rules/constant.rs#L11

partial evaluator here: https://github.com/conjure-cp/conjure-oxide/blob/dc7f98bd5cdf32122de3ce4fb8235d86747d4b88/crates/conjure_core/src/rules/partial_eval.rs

ozgurakgun replied:
and if it all was working correctly this would already do it? https://github.com/conjure-cp/conjure-oxide/blob/dc7f98bd5cdf32122de3ce4fb8235d86747d4b88/crates/conjure_core/src/rules/constant.rs#L42

I think there is a bug. how well tested is the full evaluator?

remove_empty_expression is at 8800, and full evaluator is at 9001, so that should take precedence even if this rule was buggy (which I think it is)

niklasdewally commented:
@YehorBoiar I've just merged some code that checks solutions against Conjure when running ACCEPT=true cargo test. 
It might be helpful to update your branch to main and rerun ACCEPT=true cargo test to make sure all this still works.

(https://github.com/conjure-cp/conjure-oxide/commit/0c9118b831c3aa0ab131d5c9a9a4fa25f673c26f)

YehorBoiar commented:
Okay, thanks.

ozgurakgun commented on conjure_oxide/tests/generated_tests.rs:
in addition to removing nothing you introduce an `if let` here, why is that?

ozgurakgun commented on crates/conjure_core/src/rule_engine/rewrite.rs:
```diff
@@ -130,10 +132,13 @@ pub fn rewrite_model<'a>(
 
     let start = std::time::Instant::now();
-
+    let default_constraint = Expression::Constant(Metadata::new(), Constant::Int(1));
```
when the top level is a vector we won't need magic values like this one...

ozgurakgun commented on crates/conjure_core/src/rules/base.rs:
```diff
@@ -21,90 +20,61 @@ register_rule_set!("Base", 100, ());
-
-    Ok(Reduction::pure(new_expr))
-}
+// #[register_rule(("Base", 8800))]
```
don't comment out code please, just remove it!

niklasdewally commented on conjure_oxide/tests/integration/basic/bool/02/bool-02.expected-rewrite.serialised.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "constraints": "Nothing",
+  "constraints": null,
```
**FYI:** Given that this passes, it seems that null passes through to Minion without erroring: ideally, the Minion solver adaptor shouldn't support `null` at all. 

This is possibly outside the scope of this PR, given that you did not introduce this change.

YehorBoiar commented:
@ozgurakgun @niklasdewally I'm getting this error on rules now, but I don't understand why. Could you help?

```
mismatched types
expected fn pointer `for<'a, 'b> fn(&'a std::vec::Vec<Expression>, &'b model::Model) -> std::result::Result<_, _>`
      found fn item `for<'a, 'b> fn(&'a Expression, &'b model::Model) -> std::result::Result<_, _> {unwrap_sum}`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[2]?2#file:///cs/home/yb33/Documents/VIP/conjure-oxide/crates/conjure_core/src/rules/base.rs)
```

ozgurakgun commented on crates/conjure_core/src/rule_engine/rule.rs:
```diff
@@ -63,7 +63,11 @@ pub struct Reduction {
 impl Reduction {
-    pub fn new(new_expression: Expression, new_top: Expression, symbols: SymbolTable) -> Self {
+    pub fn new(
+        new_expression: Vec<Expression>,
```
@YehorBoiar this shouldn't be a vec. it's the replacement expression, so you want to replace expression with expression. this is probably the reason for the error you are getting.

ozgurakgun commented on crates/conjure_core/src/rule_engine/rule.rs:
```diff
@@ -72,25 +76,25 @@ impl Reduction {
 
     /// Represents a reduction that also modifies the symbol table.
-    pub fn with_symbols(new_expression: Expression, symbols: SymbolTable) -> Self {
+    pub fn with_symbols(new_expression: Vec<Expression>, symbols: SymbolTable) -> Self {
```
propagate that change here too. we just want to change the constraints in the model (and new_top as well) to be a vec, not everything else.

ozgurakgun commented:
just to make sure this is not blocked on me @YehorBoiar ?

YehorBoiar commented:
@ozgurakgun No, it's not. I was working on it, but got stuck at some point. I decided to take one day off this task and focus on gen_reduce for a little bit. I will start working on it either today's evening or tomorrow morning. I will write more details about the issue in case if I wouldn't resolve it. 

Thanks for double checking on me!

YehorBoiar commented:
When I'm running `cargo run` or `cargo test` I'm getting

```
thread 'coordinator' panicked at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/compiler/rustc_codegen_ssa/src/back/write.rs:1676:29:
/rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/compiler/rustc_codegen_ssa/src/back/write.rs:1676:29: worker thread panicked
stack backtrace:
   0:     0x7f7025bce6ea - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h304520fd6a30aa07
   1:     0x7f7026419525 - core::fmt::write::hf5713710ce10ff22
   2:     0x7f7027234d91 - std::io::Write::write_fmt::hda708db57927dacf
   3:     0x7f7025bd0dbb - std::panicking::default_hook::{{closure}}::he1ad87607d0c11c5
   4:     0x7f7025bd0a2e - std::panicking::default_hook::h81c8cd2e7c59ee33
   5:     0x7f7024d9a5d7 - std[5204e9590b4985ef]::panicking::update_hook::<alloc[fd15fd9026f491e1]::boxed::Box<rustc_driver_impl[c41f2638408ed175]::install_ice_hook::{closure#0}>>::{closure#0}
   6:     0x7f7025bd16d7 - std::panicking::rust_panic_with_hook::had2118629c312a4a
   7:     0x7f70252f2892 - std[5204e9590b4985ef]::panicking::begin_panic::<alloc[fd15fd9026f491e1]::string::String>::{closure#0}
   8:     0x7f70252ebb76 - std[5204e9590b4985ef]::sys::backtrace::__rust_end_short_backtrace::<std[5204e9590b4985ef]::panicking::begin_panic<alloc[fd15fd9026f491e1]::string::String>::{closure#0}, !>
   9:     0x7f70252dec11 - std[5204e9590b4985ef]::panicking::begin_panic::<alloc[fd15fd9026f491e1]::string::String>
  10:     0x7f70253760bd - rustc_middle[c83967c7761a8780]::util::bug::opt_span_bug_fmt::<rustc_span[233999951ced9cd1]::span_encoding::Span>::{closure#0}
  11:     0x7f702535b11a - rustc_middle[c83967c7761a8780]::ty::context::tls::with_opt::<rustc_middle[c83967c7761a8780]::util::bug::opt_span_bug_fmt<rustc_span[233999951ced9cd1]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  12:     0x7f702535afcb - rustc_middle[c83967c7761a8780]::ty::context::tls::with_context_opt::<rustc_middle[c83967c7761a8780]::ty::context::tls::with_opt<rustc_middle[c83967c7761a8780]::util::bug::opt_span_bug_fmt<rustc_span[233999951ced9cd1]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  13:     0x7f7022be2960 - rustc_middle[c83967c7761a8780]::util::bug::bug_fmt
  14:     0x7f7027273a1a - rustc_codegen_ssa[4b0e0146219e1624]::back::write::start_executing_work::<rustc_codegen_llvm[1041eb84fe8a92c6]::LlvmCodegenBackend>::{closure#5}
  15:     0x7f70272714f7 - std[5204e9590b4985ef]::sys::backtrace::__rust_begin_short_backtrace::<<rustc_codegen_llvm[1041eb84fe8a92c6]::LlvmCodegenBackend as rustc_codegen_ssa[4b0e0146219e1624]::traits::backend::ExtraBackendMethods>::spawn_named_thread<rustc_codegen_ssa[4b0e0146219e1624]::back::write::start_executing_work<rustc_codegen_llvm[1041eb84fe8a92c6]::LlvmCodegenBackend>::{closure#5}, core[d89802b8f5f07590]::result::Result<rustc_codegen_ssa[4b0e0146219e1624]::back::write::CompiledModules, ()>>::{closure#0}, core[d89802b8f5f07590]::result::Result<rustc_codegen_ssa[4b0e0146219e1624]::back::write::CompiledModules, ()>>
  16:     0x7f702727773b - <<std[5204e9590b4985ef]::thread::Builder>::spawn_unchecked_<<rustc_codegen_llvm[1041eb84fe8a92c6]::LlvmCodegenBackend as rustc_codegen_ssa[4b0e0146219e1624]::traits::backend::ExtraBackendMethods>::spawn_named_thread<rustc_codegen_ssa[4b0e0146219e1624]::back::write::start_executing_work<rustc_codegen_llvm[1041eb84fe8a92c6]::LlvmCodegenBackend>::{closure#5}, core[d89802b8f5f07590]::result::Result<rustc_codegen_ssa[4b0e0146219e1624]::back::write::CompiledModules, ()>>::{closure#0}, core[d89802b8f5f07590]::result::Result<rustc_codegen_ssa[4b0e0146219e1624]::back::write::CompiledModules, ()>>::{closure#1} as core[d89802b8f5f07590]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  17:     0x7f7027277d2b - std::sys::pal::unix::thread::Thread::new::thread_start::hcdbd1049068002f4
  18:     0x7f7021689c02 - start_thread
                               at /usr/src/debug/glibc-2.34-100.el9_4.4.x86_64/nptl/pthread_create.c:443:8
  19:     0x7f702170ec40 - clone3
                               at /usr/src/debug/glibc-2.34-100.el9_4.4.x86_64/misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  20:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.82.0 (f6e511eec 2024-10-15) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
```
I'm not sure what caused it, and I don't know how to navigate this error message. @ozgurakgun @niklasdewally Do you have any idea how I can fix it?

niklasdewally commented:
> error: the compiler unexpectedly panicked. this is a bug.

> note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

> note: rustc 1.82.0 (f6e511eec 2024-10-15) running on x86_64-unknown-linux-gnu


Seems like a rust bug, nothing to do with you or us....

Maybe try cargo clean and updating rust? Failing that, sending Rust a big report (with the link above) would be the best move...

YehorBoiar commented:
Thanks!

YehorBoiar commented:
Hey! I forgot how to generate a parse, rewrite models when testing essence. Can somebody remind me how to do that? @niklasdewally @Soph1514

niklasdewally commented:
> Hey! I forgot how to generate a parse, rewrite models when testing essence. Can somebody remind me how to do that? @niklasdewally @Soph1514 

ACCEPT=true cargo test

YehorBoiar commented:
I'm getting in an infinte loop when running tests. 
![image](https://github.com/user-attachments/assets/b036d906-1daa-4a61-8bfe-2f46e05485df)
Is it because my apply function behaves funny?

niklasdewally commented:
What do the logs say?

YehorBoiar commented:
I'm getting failures on 

```

failures:

---- tests_integration_eprime_minion_partial_eval_add stdout ----
WARNING: Unimplemented constant eval: SumEq(Metadata { clean: false, etype: None }, [Reference(Metadata { clean: false, etype: None }, UserName("x")), Constant(Metadata { clean: false, etype: None }, Int(10)), Reference(Metadata { clean: false, etype: None }, UserName("y")), Constant(Metadata { clean: false, etype: None }, Int(25))], Constant(Metadata { clean: false, etype: None }, Int(100)))
WARNING: Unimplemented constant eval: SumEq(Metadata { clean: false, etype: None }, [Reference(Metadata { clean: false, etype: None }, UserName("x")), Reference(Metadata { clean: false, etype: None }, UserName("y"))], Constant(Metadata { clean: false, etype: None }, Int(65)))
Building Minion model...
Running Minion...
Error: ConjureSolutionsError("expected value at line 6 column 1")

---- tests_integration_eprime_minion_partial_eval_01_add stdout ----
WARNING: Unimplemented constant eval: SumEq(Metadata { clean: false, etype: None }, [Reference(Metadata { clean: false, etype: None }, UserName("x")), Constant(Metadata { clean: false, etype: None }, Int(10)), Reference(Metadata { clean: false, etype: None }, UserName("y")), Constant(Metadata { clean: false, etype: None }, Int(25))], Constant(Metadata { clean: false, etype: None }, Int(100)))
WARNING: Unimplemented constant eval: SumEq(Metadata { clean: false, etype: None }, [Reference(Metadata { clean: false, etype: None }, UserName("x")), Reference(Metadata { clean: false, etype: None }, UserName("y"))], Constant(Metadata { clean: false, etype: None }, Int(65)))
Building Minion model...
Running Minion...
Error: ConjureSolutionsError("expected value at line 6 column 1")

```

I suspect that the reason for that is the fact that I deleted the rule that removes nothings. However, I tried to rewrite it to this, and it only made things worse (even less tests were passing and the results were chaotic)

```
/// Remove empty Ands from expressions:
/// ```text
/// and([a, And([], ...), b]) = and([a, b])
/// sum([a, And([], ...), b]) = sum([a, b])
/// sum_leq([a, And([], ...), b], c) = sum_leq([a, b], c)
/// ...
/// ```
#[register_rule(("Base", 8800))]
fn remove_empty_ands(expr: &Expression, _: &Model) -> ApplicationResult {
    fn filter_empty_ands(exprs: Vec<Expression>) -> Result<Vec<Expression>, ApplicationError> {
        let mut changed = false;
        let mut new_exprs = Vec::new();

        for e in exprs {
            match e.clone() {
                Expression::And(_, ref sub_exprs) if sub_exprs.is_empty() => {
                    changed = true;
                }
                _ => new_exprs.push(e),
            }
        }

        if changed {
            Ok(new_exprs)
        } else {
            Err(ApplicationError::RuleNotApplicable)
        }
    }

    fn get_lhs_rhs(sub: Vec<Expression>) -> (Vec<Expression>, Box<Expression>) {
        if sub.is_empty() {
            return (Vec::new(), Box::new(Expression::And(Metadata::new(), Vec::new())));
        }

        let lhs = sub[..(sub.len() - 1)].to_vec();
        let rhs = Box::new(sub[sub.len() - 1].clone());
        (lhs, rhs)
    }

    let new_sub = filter_empty_ands(expr.children().into_iter().collect())?;

    match expr {
        Expression::And(md, _) => Ok(Reduction::pure(Expression::And(md.clone(), new_sub))),
        Expression::Or(md, _) => Ok(Reduction::pure(Expression::Or(md.clone(), new_sub))),
        Expression::Sum(md, _) => Ok(Reduction::pure(Expression::Sum(md.clone(), new_sub))),
        Expression::SumEq(md, _, _) => {
            let (lhs, rhs) = get_lhs_rhs(new_sub);
            Ok(Reduction::pure(Expression::SumEq(md.clone(), lhs, rhs)))
        }
        Expression::SumLeq(md, _, _) => {
            let (lhs, rhs) = get_lhs_rhs(new_sub);
            Ok(Reduction::pure(Expression::SumLeq(md.clone(), lhs, rhs)))
        }
        Expression::SumGeq(md, _, _) => {
            let (lhs, rhs) = get_lhs_rhs(new_sub);
            Ok(Reduction::pure(Expression::SumGeq(md.clone(), lhs, rhs)))
        }
        _ => Err(ApplicationError::RuleNotApplicable),
    }
}
```

@niklasdewally Do you think that it is the reason why we currently fail, or it could be something else in my implementation that went wrong?

niklasdewally commented:
> Error: ConjureSolutionsError("expected value at line 6 column 1")

This is something to do with reading the solutions against json output from Conjure, which happens after we run the model.

Try to `conjure solve` the failing test manually, and see what happens.
Otherwise, we can stick this in a debugger and see exactly where this error comes from.

niklasdewally commented:
If you still suspect a rule issue, it might help to update your branch against main, as I have recently improved our logging output:

```
 RUST_LOG=TRACE cargo run -- --verbose conjure-oxide/<path>/<to>/input.essence
```

YehorBoiar commented:
Sorry, for bother, I run the same script on my own machine, and it worked just fine. All tests are being passed. Please review @ozgurakgun

ozgurakgun commented on crates/conjure_core/src/rule_engine/rule.rs:
```diff
@@ -101,21 +101,24 @@ impl Reduction {
-                    vec![self.new_expression.clone(), self.new_top],
-                ),
-            };
+        if let Expression::And(_, exprs) = &self.new_top {
```
I would add a TODO comment here about the hacky and([]) use/check. We shouldn't forget to remove it when we change the constraints to a vec.

ozgurakgun commented on crates/conjure_core/src/rules/constant.rs:
```diff
@@ -4,14 +4,23 @@ use conjure_core::rule_engine::{
 
 #[register_rule(("Constant", 9001))]
 fn apply_eval_constant(expr: &Expr, _: &Model) -> ApplicationResult {
+    // Check for expressions excluded by remove_empty_expression
```
again, why is this change in this PR? it looks unrelated?

YehorBoiar replied:
When I started doing this PR we were talking that this rule is unnecessary, so I decided to remove it. However, you're right, this change doesn't belong to this PR. I would do it in a separate one.

niklasdewally reviewed (commented):
@YehorBoiar you should update this branch to main; there are some merge conflicts due to  https://github.com/conjure-cp/conjure-oxide/commit/f931a3d2b407c5f295e7eab181d53bd26a32c6aa

YehorBoiar commented:
@niklasdewally Cheers, done!
